### PR TITLE
Chunter: Channel attributes

### LINKED ISSUE #1248: Chunter: Channel attributes

### PR DESCRIPTION
**Pull request by ksilobait**
https://github.com/hcengineering/anticrm/issues/1248

### CODE REVIEW

File: models/chunter/src/index.ts [hunk 1]
```diff
@@ -34,6 +34,11 @@ export const DOMAIN_COMMENT = 'comment' as Domain
 export class TChannel extends TSpace implements Channel {
   @Prop(TypeTimestamp(), chunter.string.LastMessage)
   lastMessage?: Timestamp
+
+  // TODO: check if this is correct (FullText/Indexed)
+  @Prop(TypeMarkup(), chunter.string.Topic)
```
Comment (BykhovDenis) [inline models/chunter/src/index.ts]: TypeString() here if you use EditBox for edit, or TypeMarkup() if you use something like StyledTextBox

File: packages/presentation/src/components/Members.svelte [hunk 2]
```diff
@@ -0,0 +1,36 @@
+<!--
+// Copyright © 2022 Anticrm Platform Contributors.
```
Comment (BykhovDenis) [inline packages/presentation/src/components/Members.svelte]: Please use `Copyright © 2022 Hardcore Engineering Inc.`

File: plugins/chunter-resources/src/components/CreateChannel.svelte [hunk 3]
```diff
@@ -25,16 +24,18 @@
...
     return name === ''
   }
   const client = getClient()
 
-  function createChannel() {
+  function createChannel () {
     client.createDoc(chunter.class.Channel, core.space.Space, {
       name,
       description,
+      topic,
```
Comment (BykhovDenis) [inline plugins/chunter-resources/src/components/CreateChannel.svelte]: Let it be `undefined`

File: plugins/chunter-resources/src/components/EditChannel.svelte [hunk 4]
```diff
@@ -0,0 +1,144 @@
+<!--
+//
+// Copyright © 2022 Anticrm Platform Contributors.
```
Comment (BykhovDenis) [inline plugins/chunter-resources/src/components/EditChannel.svelte]: `Copyright © 2022 Hardcore Engineering Inc.`

File: plugins/chunter-resources/src/components/Header.svelte [hunk 5]
```diff
@@ -0,0 +1,39 @@
+<!--
+// Copyright © 2020 Anticrm Platform Contributors.
```
Comment (BykhovDenis) [inline plugins/chunter-resources/src/components/Header.svelte]: `Copyright © 2022 Hardcore Engineering Inc.`

File: plugins/chunter/src/index.ts [hunk 6]
```diff
@@ -103,7 +104,9 @@ export default plugin(chunterId, {
     Channel: '' as Ref<Class<Channel>>
   },
   space: {
-    Backlinks: '' as Ref<Space>
+    Backlinks: '' as Ref<Space>,
+    Description: '' as Ref<Space>,
```
Comment (BykhovDenis) [inline plugins/chunter/src/index.ts]: We don't need it

File: plugins/chunter/src/index.ts [hunk 7]
```diff
@@ -103,7 +104,9 @@ export default plugin(chunterId, {
...
+    Topic: '' as Ref<Space>
```
Comment (BykhovDenis) [inline plugins/chunter/src/index.ts]: We don't need it

File: plugins/chunter-resources/src/components/EditChannel.svelte [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+      // Just refresh value
+      query.query(chunter.class.Channel, { _id }, (result) => {
+        channel = result[0]
+      })
+    }
+  }
+
+  function onTopicChange (ev: Event) {
+    const newTopic = (ev.target as HTMLInputElement).value
+    client.updateDoc(_class, chunter.space.Topic, channel._id, { topic: newTopic })
```
Comment (BykhovDenis) [inline plugins/chunter-resources/src/components/EditChannel.svelte]: use channel.space as space

File: plugins/chunter-resources/src/components/EditChannel.svelte [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+  }
+
+  function onDescriptionChange (ev: Event) {
+    const newDescription = (ev.target as HTMLInputElement).value
+    client.updateDoc(_class, chunter.space.Description, channel._id, { description: newDescription })
```
Comment (BykhovDenis) [inline plugins/chunter-resources/src/components/EditChannel.svelte]: use channel.space as space

**Review by BykhovDenis — CHANGES_REQUESTED** [review]

File: plugins/chunter-resources/src/components/EditChannel.svelte (hunk 8 above)
Comment (BykhovDenis) [inline plugins/chunter-resources/src/components/EditChannel.svelte]: Also you can use `client.update(channel, { topic: newTopic })

File: plugins/chunter-resources/src/components/EditChannel.svelte [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+          />
+          <EditBox
+            label={chunter.string.ChannelDescription}
+            bind:value={channel.description}
+            placeholder={chunter.string.ChannelDescription}
+            maxWidth="39rem"
+            focus
+            on:change={onDescriptionChange}
+          />
+          <Attachments objectId={channel._id} _class={channel._class} space={channel.space} />
```
Comment (BykhovDenis) [inline plugins/chunter-resources/src/components/EditChannel.svelte]: Here we need another component to show all space attachments, table for example, please remove it currenlty

**Review by BykhovDenis — CHANGES_REQUESTED** [review]

File: plugins/chunter-resources/src/components/EditChannel.svelte (hunk 10 above)
Comment (ksilobait) [reply plugins/chunter-resources/src/components/EditChannel.svelte]: Done

File: plugins/chunter-resources/src/components/EditChannel.svelte (hunk 8 above)
Comment (ksilobait) [reply plugins/chunter-resources/src/components/EditChannel.svelte]: Done

File: plugins/chunter-resources/src/components/EditChannel.svelte (hunk 9 above)
Comment (ksilobait) [reply plugins/chunter-resources/src/components/EditChannel.svelte]: Replaced with client.update

File: plugins/chunter-resources/src/components/EditChannel.svelte (hunk 8 above)
Comment (ksilobait) [reply plugins/chunter-resources/src/components/EditChannel.svelte]: Replaced with client.update

File: plugins/chunter/src/index.ts (hunk 7 above)
Comment (ksilobait) [reply plugins/chunter/src/index.ts]: Done

File: plugins/chunter/src/index.ts (hunk 6 above)
Comment (ksilobait) [reply plugins/chunter/src/index.ts]: Done

File: plugins/chunter-resources/src/components/Header.svelte (hunk 5 above)
Comment (ksilobait) [reply plugins/chunter-resources/src/components/Header.svelte]: Done

File: plugins/chunter-resources/src/components/EditChannel.svelte (hunk 4 above)
Comment (ksilobait) [reply plugins/chunter-resources/src/components/EditChannel.svelte]: Done

File: plugins/chunter-resources/src/components/CreateChannel.svelte (hunk 3 above)
Comment (ksilobait) [reply plugins/chunter-resources/src/components/CreateChannel.svelte]: Done

File: packages/presentation/src/components/Members.svelte (hunk 2 above)
Comment (ksilobait) [reply packages/presentation/src/components/Members.svelte]: Done

File: models/chunter/src/index.ts (hunk 1 above)
Comment (ksilobait) [reply models/chunter/src/index.ts]: Done

**Review by BykhovDenis — APPROVED** [review]

File: packages/presentation/lang/en.json.orig [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+    "Ok": "Ok",
+    "Save": "Save",
+    "Download": "Download",
+    "Close": "Close",
+    "NotSelected": "Not selected",
+    "Deselect": "Deselect",
+    "AddSocialLinks": "Add social links",
+    "Change": "Change",
+    "Remove": "Remove",
+<<<<<<< HEAD
```
Comment (haiodo) [inline packages/presentation/lang/en.json.orig]: Conflict should be resolved before merge.

File: packages/presentation/lang/ru.json.orig [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+    "Ok": "Ок",
+    "Save": "Сохранить",
+    "Download": "Загрузить",
+    "Close": "Закрыть",
+    "NotSelected": "Не выбрано",
+    "Deselect": "Снять выделение",
+    "AddSocialLinks": "Добавить контактную информацию",
+    "Change": "Изменить",
+    "Remove": "Удалить",
+<<<<<<< HEAD
```
Comment (haiodo) [inline packages/presentation/lang/ru.json.orig]: Conflict should be resolved before merge.

File: packages/presentation/src/plugin.ts.orig [hunk 13]
```diff
@@ -0,0 +1,53 @@
+//
+// Copyright © 2020, 2021 Anticrm Platform Contributors.
+// Copyright © 2021 Hardcore Engineering Inc.
```
Comment (haiodo) [inline packages/presentation/src/plugin.ts.orig]: Only `Copyright © 2022 Hardcore Engineering Inc.` header is required.

File: packages/presentation/src/plugin.ts.orig [hunk 14]
```diff
@@ -0,0 +1,53 @@
+//
```
Comment (haiodo) [inline packages/presentation/src/plugin.ts.orig]: why we need this file at all?

File: plugins/chunter-resources/src/backlinks.ts [hunk 15]
```diff
@@ -23,10 +26,20 @@ export function getTime (time: number): string {
...
 }
 
-function extractBacklinks (backlinkId: Ref<Doc>, backlinkClass: Ref<Class<Doc>>, attachedDocId: Ref<Doc> | undefined, message: string, kids: NodeListOf<ChildNode>): Array<Data<Backlink>> {
+function extractBacklinks (
+  backlinkId: Ref<Doc>,
+  backlinkClass: Ref<Class<Doc>>,
+  attachedDocId: Ref<Doc> | undefined,
+  message: string,
+  kids: NodeListOf<ChildNode>
+): Array<Data<Backlink>> {
```
Comment (haiodo) [inline plugins/chunter-resources/src/backlinks.ts:42]: Why not ?
```suggestion
): Data<Backlink>[] {
```

File: plugins/chunter-resources/src/components/ChannelHeader.svelte [hunk 16]
```diff
@@ -0,0 +1,49 @@
+<!--
+// Copyright © 2020 Anticrm Platform Contributors.
```
Comment (haiodo) [inline plugins/chunter-resources/src/components/ChannelHeader.svelte:2]: Copyright © 2022 Hardcore Engineering Inc.

Comment (haiodo) [inline plugins/chunter-resources/src/components/ChannelHeader.svelte]:
```suggestion
// Copyright © 2022 Hardcore Engineering Inc.
```

File: plugins/chunter-resources/src/components/CreateChannel.svelte [hunk 17]
```diff
@@ -25,13 +24,13 @@
   const dispatch = createEventDispatcher()
 
   let name: string = ''
-  let description: string = ''
-  export function canClose(): boolean {
+  const description: string = ''
```
Comment (haiodo) [inline plugins/chunter-resources/src/components/CreateChannel.svelte]: Please add description editor control or remove it at all.

File: plugins/chunter-resources/src/components/EditChannel.svelte [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  function onDescriptionChange (ev: Event) {
+    const newDescription = (ev.target as HTMLInputElement).value
+    client.update(channel, { description: newDescription })
+  }
+</script>
+
+<div
+  class="antiOverlay"
```
Comment (haiodo) [inline plugins/chunter-resources/src/components/EditChannel.svelte]: This is redundant if used inside showPopup or showPanel.

File: plugins/chunter-resources/src/components/EditChannel.svelte [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+      >
+        <Label label={tabLabel} />
+      </div>
+    {/each}
+    <div class="ac-tabs__empty" />
+  </div>
+  <Scroller padding>
+    {#if selectedTabIndex === 0}
+      <!-- Channel description -->
+      {#if channel}
```
Comment (haiodo) [inline plugins/chunter-resources/src/components/EditChannel.svelte]: Better to move every tab into separate svelte component.

File: packages/presentation/lang/en.json.orig (hunk 11 above)
Comment (ksilobait) [reply packages/presentation/lang/en.json.orig]: Removed this file (conflict is already resolved, just forgot not to commit this file)

File: packages/presentation/lang/ru.json.orig (hunk 12 above)
Comment (ksilobait) [reply packages/presentation/lang/ru.json.orig]: > removed this file

File: packages/presentation/src/plugin.ts.orig (hunk 14 above)
Comment (ksilobait) [reply packages/presentation/src/plugin.ts.orig]: > removed this file

File: packages/presentation/src/plugin.ts.orig (hunk 13 above)
Comment (ksilobait) [reply packages/presentation/src/plugin.ts.orig]: > removed this file

Should I update `packages/presentation/src/plugin.ts` as well?

File: plugins/chunter-resources/src/backlinks.ts (hunk 15 above)
Comment (ksilobait) [reply plugins/chunter-resources/src/backlinks.ts:42]: I think it's not related to this PR but I've updated this

File: plugins/chunter-resources/src/components/ChannelHeader.svelte (hunk 16 above)
Comment (ksilobait) [reply plugins/chunter-resources/src/components/ChannelHeader.svelte]: Done

File: plugins/chunter-resources/src/components/EditChannel.svelte (hunk 18 above)
Comment (ksilobait) [reply plugins/chunter-resources/src/components/EditChannel.svelte]: So I should remove this for `PDFViewer`, EditStatuses (but not for EditVacancy as there's a href), right?

File: plugins/chunter-resources/src/components/EditChannel.svelte (hunk 19 above)
Comment (ksilobait) [reply plugins/chunter-resources/src/components/EditChannel.svelte]: Extracted Description

File: plugins/chunter-resources/src/components/CreateChannel.svelte (hunk 17 above)
Comment (ksilobait) [reply plugins/chunter-resources/src/components/CreateChannel.svelte]: Inlined this string

File: plugins/chunter-resources/src/backlinks.ts (hunk 15 above)
Comment (ksilobait) [reply plugins/chunter-resources/src/backlinks.ts:42]: Either way `npm run format` converts this back to `Array<...>`

**Review by haiodo — APPROVED** [review]

File: plugins/chunter-resources/src/components/EditChannel.svelte (hunk 18 above)
Comment (haiodo) [reply plugins/chunter-resources/src/components/EditChannel.svelte]: Please update only files related to current task, and create issues for improvements for further planing.

**Pull request merged**